batch_requests.go

```diff
@@ -0,0 +1,150 @@
+	}
+
+	br.Requests = append(br.Requests, batchItem)
+	return batchItem, nil
+}
+
+// DependsOnItem creates a dependency chain between BatchItems.If A depends on B, then B will be sent before B
+func (bi *batchItem) DependsOnItem(item batchItem) {
+	// DependsOn is a single value slice
+	bi.DependsOn = []string{item.Id}
+}
+
+// NewBatchRequest creates an instance of BatchRequest
+func NewBatchRequest() *batchRequest {
+	return &batchRequest{}
+}
+
+func toBatchItem(requestInfo abstractions.RequestInformation) (*batchItem, error) {
+	url, err := requestInfo.GetUri()
+	if weGotAnError(err) {
+		return nil, err
+	}
+
+	var body map[string]any
+	json.Unmarshal(requestInfo.Content, &body)
+
+	return &batchItem{
+		Id:        uuid.NewString(),
+		Method:    requestInfo.Method.String(),
+		Body:      body,
+		Headers:   requestInfo.Headers,
+		Url:       url.Path,
+		DependsOn: make([]string, 0),
+	}, nil
+}
+
+func (r batchRequest) toJson() ([]byte, error) {
+	return json.Marshal(r)
+}
+
+func getBaseUrl(adapter abstractions.RequestAdapter) (*url.URL, error) {
+	return url.Parse(adapter.GetBaseUrl())
+}
+
+func buildRequestInfo(jsonBody []byte, baseUrl *url.URL) *abstractions.RequestInformation {
+	requestInfo := abstractions.NewRequestInformation()
+
+	requestInfo.SetStreamContent(jsonBody)
+	requestInfo.Method = abstractions.POST
+	requestInfo.SetUri(*baseUrl)
+	requestInfo.Headers = map[string]string{
+		"Content-Type": "application/json",
+	}
+
+	return requestInfo
+}
+
+func sendBatchRequest(requestInfo *abstractions.RequestInformation, adapter abstractions.RequestAdapter) (BatchResponse, error) {
+	var res BatchResponse
+
+	// we don't care about SendAsync's return value because we don't use it and it is a workaround
+	// on SendAsync's insistence to have a parsable node as a return value
+	_, err := adapter.SendAsync(requestInfo, nil, func(response any, errorMappings abstractions.ErrorMappings) (any, error) {
+		resp, ok := response.(*http.Response)
+		if !ok {
+			return nil, errors.New("Response type assertion failed")
+		}
+
+		if status := resp.StatusCode; status >= 400 && status <= 600 {
+			return res, fmt.Errorf("Request failed with status: %d", status)
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if weGotAnError(err) {
+			return res, err
+		}
+
+		json.Unmarshal(body, &res)
+
+		// returning a Noop Parsable here because SendAsync type asserts HandlerFunc's return value
+		return internal.NewNoOpParsable(), nil
+	}, nil)
+
+	if weGotAnError(err) {
+		return res, err
+	}
+
+	return res, nil
+}
+
+type batchRequest struct {
+	Requests []*batchItem `json:"requests"`
+}
+
+type BatchResponse struct {
+	Responses []batchItem
+}
+
+type batchItem struct {
+	Id        string            `json:"id"`
+	Method    string            `json:"method"`
+	Url       string            `json:"url"`
+	Headers   map[string]string `json:"headers"`
+	Body      map[string]any    `json:"body"`
+	DependsOn []string          `json:"dependsOn"`
+}
```

Comment:
how are people supposed to deserialize to a model?

Reply:
Serialise what to a model?

Reply:
the response body

Reply:
Users are going to unmarshal the body into a struct.

Reply:
The value of body is a raw json string. Returning a parsable here leaks the abstraction and it will be a less than ideal scenario for the user.

Reply:
Right, could we then have a `GetResponseById(ParsableFactory)` so people have access to a deserialized model quickly?
We might need declinations of it for scalar and collections values as well, unless you want to explore the new generics? (which we're planning to leverage to reduce the amount of code generated for service libs).
This way the batching implementation here will provide the same level of service as the other languages.
https://docs.microsoft.com/en-us/graph/sdks/batch-requests?tabs=csharp

Reply:
You'll need to pass the factory to handle discriminator (odata type) for parsable models, but maybe that parameter is optional so the scalar types and collections of scalar types don't need to provide it